### PR TITLE
Verify SyntaxList length in SyntaxNodeCache.TryGetNode()

### PR DESCRIPTION
**Pull request by cston**
Verify `node.SlotCount` matches expected length in `SyntaxNodeCache.TryGetNode()`.

Fixes infrequent `InvalidCastException` in [`SyntaxList.List()`](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxList.cs#L35).

### CODE REVIEW

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeCacheTests.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+                var listOf3 = new Microsoft.CodeAnalysis.Syntax.InternalSyntax.SyntaxList.WithThreeChildren(child0, child1, child2);
+                int hashOf3 = listOf3.GetCacheHash();
+                SyntaxNodeCache.AddNode(listOf3, hashOf3);
+
+                var listCached = SyntaxNodeCache.TryGetNode(listOf3.RawKind, child0, child1, child2, SyntaxNodeCache.GetDefaultNodeFlags(), out _);
+                Assert.NotNull(listCached);
+
+                var listOf2 = SyntaxNodeCache.TryGetNode(listOf3.RawKind, child0, child1, SyntaxNodeCache.GetDefaultNodeFlags(), out _);
+                Assert.True(listOf3 != listOf2, $"{i} iterations");
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeCacheTests.cs:64]: With `net8.0`, this check failed within ~500M iterations.

**Review by CyrusNajmabadi — APPROVED** [review]

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeCacheTests.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable disable
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeCacheTests.cs:5]: Nit: why disable this new test file?

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeCacheTests.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeCacheTests.cs:5]: For tests, in general, it seems useful to allow nullable mismatches. That said, I don't know if there any such cases in this test file currently.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeCacheTests.cs:5]: I also do not find nullable analysis in unit-test files useful. To be precise, I even find it distracting, i.e. getting in a way.

**Pull request merged**